src/Launcher.vala

```diff
@@ -259,9 +261,13 @@ public class Dock.Launcher : Gtk.Box {
         app.notify["running"].connect (update_running_revealer);
         update_running_revealer ();
 
+        app.notify["running"].connect(close_all_instances);
```

Comment:
All of this should be done in the App class.
Also `close_all_instances` sounds like the method will actually close all instances. Something like `update_close_action` could be more fitting IMHO